### PR TITLE
Implemented :class:`~.ImplicitFunction` and :meth:`.CoordinateSystem.get_implicit_curve` for plotting implicit curves

### PR DESCRIPTION
**Pull request by icedcoffeeee**
## Overview: What does this pull request change?

* Added new :class:`~.ImplicitFunction` mobject 
* Added new :meth:`.CoordinateSystem.get_implicit_graph`

This is a direct continuation of #2046 , my fork went missing.
## Reviewer Checklist
- [x] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [x] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [x] If applicable: newly added functions and classes are tested

### CODE REVIEW

File: manim/mobject/coordinate_systems.py [hunk 1]
```diff
@@ -697,6 +697,58 @@ def log_func(x):
...
+        Parameters
+        ----------
+        func
+            The function to graph, in the form of f(x, y) = 0.
+        min_depth
+            The minimum depth of the function to calculate.
+        max_quads
+            The maximum number of quads to use.
+        color
+            The color of the graph
```
Comment (Darylgolden) [inline manim/mobject/coordinate_systems.py:719]:
```suggestion
            The color of the graph.
```

File: manim/mobject/functions.py [hunk 2]
```diff
@@ -147,3 +152,80 @@ def get_function(self):
...
+            np.array([self.x_range[1], self.y_range[1]]),
+        )
+        points = plot_isoline(
+            lambda u: self.function(u[0], u[1]),
+            p_min,
+            p_max,
+            self.min_depth,
+            self.max_quads,
+        )
+        points = [np.pad(i, [(0, 0), (0, 1)]) for i in points if i != []]
```
Comment (Darylgolden) [inline manim/mobject/functions.py:223]: Perhaps `point_lists` is a better variable name?

File: manim/mobject/functions.py [hunk 3]
```diff
@@ -147,3 +152,80 @@ def get_function(self):
...
+        for p in points:
```
Comment (Darylgolden) [inline manim/mobject/functions.py:224]: I think "for list in point_lists" is more descriptive.

File: manim/mobject/functions.py [hunk 4]
```diff
@@ -1,9 +1,14 @@
...
-__all__ = ["ParametricFunction", "FunctionGraph"]
+__all__ = ["ParametricFunction", "FunctionGraph", "ImplicitFunction"]
 
 
+from typing import Callable, Optional, Sequence
+
 import numpy as np
+from isosurfaces import plot_isoline
+
+from manim.mobject.geometry import Line
```
Comment (Darylgolden) [inline manim/mobject/functions.py:11]: This is now unused, please remove.

**Review by Darylgolden — CHANGES_REQUESTED** [review]
Some more suggestions, in addition to the ones in the previous PR.

**Review by Darylgolden — APPROVED** [review]
LGTM! Maybe the number of quads could use some tuning depending on the whether we want to prioritize speed or accuracy, but that's just a minor point.

**Comment by PhotonSpheres** [comment]
A small nitpick of mine would be that `graph` doesn't seem to be the accurate term, no? I think this would be called curve or something similar. But I understand that graph might be more user-friendly.

**Comment by icedcoffeeee** [comment]
> I think this would be called curve or something similar. 

@PhotonSpheres 
I don' t mind changing the method name to `.get_implicit_curve` is that's preferred.

**Comment by PhotonSpheres** [comment]
> @PhotonSpheres I don't mind changing the method name to `.get_implicit_curve` if that's preferred.

Not sure if I even prefer that, just wanted to mention that.

**Comment by Darylgolden** [comment]
If there's no clear preference, I think it's up to you @icedcoffeeee

File: manim/mobject/coordinate_systems.py [hunk 5]
```diff
@@ -697,6 +697,58 @@ def log_func(x):
         graph.underlying_function = function
         return graph
 
+    def get_implicit_graph(
```
Comment (Darylgolden) [inline manim/mobject/coordinate_systems.py:700]: `use_smoothing` should be a parameter here too.

**Review by Darylgolden — CHANGES_REQUESTED** [review]

**Comment by Darylgolden** [comment]
I tried changing the `min_depth` and `max_quads` and there doesn't seem to be any noticeable difference in quality or speed, strange...

**Comment by icedcoffeeee** [comment]
> there's no clear preference, I think it's up to you

After googling the term implicit graph, which brings up something totally different, I'll change it to curve.

> changing the min_depth and max_quads and there doesn't seem to be any noticeable difference in quality or speed

Ah, that's something I forgot to mention in the docs. The depth is technically also the number of quads, where quads = 2 ^ depth. The higher value between them is taken into calculation.

**Comment by behackl** [comment]
> If there's no clear preference, I think it's up to you @icedcoffeeee

Given that creating a "normal plot" is done via `get_graph`, one could make an argument for `get_implicit_graph`, but generally I also prefer `get_implicit_curve` or `get_implicit_plot`.

(Actually I also don't like that all of these start with `get`, but well, that is another story.)

I think it is fine the way it is now, anyways.

**Review by Darylgolden — APPROVED** [review]

File: manim/mobject/functions.py [hunk 6]
```diff
@@ -147,3 +150,94 @@ def get_function(self):
...
+        min_depth
+            The minimum depth of the function to calculate.
+        max_quads
+            The maximum number of quads to use.
+        use_smoothing
+            Whether or not to smoothen the curves.
+
+
+        .. note::
+            A small min depth means that some small details might
```
Comment (behackl) [inline manim/mobject/functions.py:185]:
```suggestion
            A small ``min_depth`` :math:`d` means that some small details might
```

File: manim/mobject/functions.py [hunk 7]
```diff
@@ -147,3 +150,94 @@ def get_function(self):
...
+            be ignored if they don't cross an edge of one of the
+            (4 ^ min depth) uniform quads.
```
Comment (behackl) [inline manim/mobject/functions.py:187]:
```suggestion
            :math:`4^d` uniform quads.
```

File: manim/mobject/functions.py [hunk 8]
```diff
@@ -147,3 +150,94 @@ def get_function(self):
...
+
+            The value of max quads strongly correspond to the
```
Comment (behackl) [inline manim/mobject/functions.py:189]:
```suggestion
            The value of ``max_quads`` strongly corresponds to the
```

File: manim/mobject/functions.py [hunk 9]
```diff
@@ -147,3 +150,94 @@ def get_function(self):
...
+        p_min, p_max = (
+            np.array([self.x_range[0], self.y_range[0]]),
+            np.array([self.x_range[1], self.y_range[1]]),
+        )
+        curves = plot_isoline(
+            lambda u: self.function(u[0], u[1]),
+            p_min,
+            p_max,
+            self.min_depth,
+            self.max_quads,
```
Comment (behackl) [inline manim/mobject/functions.py:231]: As above: please pass as keyword arguments.

**Review by behackl — CHANGES_REQUESTED** [review]
LGTM, only have very minor change requests.

**Review by hydrobeam — COMMENTED** [review]
This PR pretty much LGTM, but I left some nitpicky comments.

Although, testing with the opengl renderer, there's some weird artefacting going on, at least on my machine. I've attached videos with the two types of shaders. Projection shaders don't seem to have the same issue.

Default shaders:
https://user-images.githubusercontent.com/70682032/135565364-eaebc7e7-3e06-4b58-9134-883f3325ca86.mp4

Projection Shaders:
https://user-images.githubusercontent.com/70682032/135565289-14b6cb36-2c38-47bf-bb1f-cf555bbc639a.mp4

File: manim/mobject/coordinate_systems.py [hunk 10]
```diff
@@ -697,6 +697,62 @@ def log_func(x):
         graph.underlying_function = function
         return graph
 
+    def get_implicit_curve(
+        self,
+        func: Callable,
+        min_depth: int = 5,
+        max_quads: int = 1500,
+        use_smoothing: bool = True,
+        color: Color = WHITE,
```
[the remark this answers is not in the document]
Comment (icedcoffeeee) [reply manim/mobject/coordinate_systems.py:706]: It was initially pointed out by @Darylgolden in https://github.com/ManimCommunity/manim/pull/2094#discussion_r718086821
But if the idea is to make it more parallel with `.get_graph` implementation, should min_depth and max_quads be omitted too?

**Comment by icedcoffeeee** [comment]
> testing with the opengl renderer, there's some weird artefacting going on, at least on my machine

This is very odd. I don't get any artifacts at all with both shaders.

https://user-images.githubusercontent.com/83535735/135569903-63bd8747-6094-40d9-8ff8-7368a4c6809a.mp4

Maybe it's gfx card specific, but that doesn't make perfect sense.

**Comment by icedcoffeeee** [comment]
@hydrobeam Apparently the neater import caused a timeout for the docs. I'll revert the commit for now.

**Comment by behackl** [comment]
> @hydrobeam Apparently the neater import caused a timeout for the docs. I'll revert the commit for now.

Not sure this is related.

**Comment by icedcoffeeee** [comment]
> Not sure this is related.

Oh? I'm not sure what else.. 🤔

**Review by behackl — APPROVED** [review]
The changes look good to me, thank you! There is a minor change conflict, could you resolve it and then run `poetry lock` and commit the resulting changes to the lockfile?

**Pull request merged**